src/main/java/org/spongepowered/asm/mixin/transformer/MemberRef.java

```diff
@@ -0,0 +1,321 @@
+/*
+ * This file is part of Mixin, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.asm.mixin.transformer;
+
+import org.spongepowered.asm.lib.Handle;
+import org.spongepowered.asm.lib.Opcodes;
+import org.spongepowered.asm.lib.tree.FieldInsnNode;
+import org.spongepowered.asm.lib.tree.MethodInsnNode;
+import org.spongepowered.asm.util.ASMHelper;
+
+/**
+ * Reference to a field or method that also includes invocation
+ * instructions.
+ *
+ * To instances are defined to be equal if they both refer to the same method
+ * and have the same invocation instructions.
+ */
+public abstract class MemberRef {
+
+    /**
+     * A static reference to a method backed by an invoke instruction
+     */
+    public static final class StaticMethodRef extends MemberRef {
```

Comment:
Sorry to be nit-picky, but could you rename the inner classes to just `Method`, `Field`, `Handle`, this will make their usage idiomatically `MemberRef.Field`, etc. which I think is cleaner.